In [2]:
import pandas as pd
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
dfc = pd.read_csv('calendar.csv')
dfl = pd.read_csv('listings.csv')
dfr = pd.read_csv('reviews.csv')

In [4]:
dfc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1393570 entries, 0 to 1393569
Data columns (total 4 columns):
 #   Column      Non-Null Count    Dtype 
---  ------      --------------    ----- 
 0   listing_id  1393570 non-null  int64 
 1   date        1393570 non-null  object
 2   available   1393570 non-null  object
 3   price       934542 non-null   object
dtypes: int64(1), object(3)
memory usage: 42.5+ MB


In [6]:
dfl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 92 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   id                                3818 non-null   int64  
 1   listing_url                       3818 non-null   object 
 2   scrape_id                         3818 non-null   int64  
 3   last_scraped                      3818 non-null   object 
 4   name                              3818 non-null   object 
 5   summary                           3641 non-null   object 
 6   space                             3249 non-null   object 
 7   description                       3818 non-null   object 
 8   experiences_offered               3818 non-null   object 
 9   neighborhood_overview             2786 non-null   object 
 10  notes                             2212 non-null   object 
 11  transit                           2884 non-null   object 
 12  thumbn

In [8]:
dfr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84849 entries, 0 to 84848
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   listing_id     84849 non-null  int64 
 1   id             84849 non-null  int64 
 2   date           84849 non-null  object
 3   reviewer_id    84849 non-null  int64 
 4   reviewer_name  84849 non-null  object
 5   comments       84831 non-null  object
dtypes: int64(3), object(3)
memory usage: 3.9+ MB


In [10]:
dfc.head()

listing_id        date available   price
0      241032  2016-01-04         t  $85.00
1      241032  2016-01-05         t  $85.00
2      241032  2016-01-06         f     NaN
3      241032  2016-01-07         f     NaN
4      241032  2016-01-08         f     NaN

In [12]:
dfl.head()

id                           listing_url       scrape_id last_scraped  \
0   241032   https://www.airbnb.com/rooms/241032  20160104002432   2016-01-04   
1   953595   https://www.airbnb.com/rooms/953595  20160104002432   2016-01-04   
2  3308979  https://www.airbnb.com/rooms/3308979  20160104002432   2016-01-04   
3  7421966  https://www.airbnb.com/rooms/7421966  20160104002432   2016-01-04   
4   278830   https://www.airbnb.com/rooms/278830  20160104002432   2016-01-04   

                                  name  \
0         Stylish Queen Anne Apartment   
1   Bright & Airy Queen Anne Apartment   
2  New Modern House-Amazing water view   
3                   Queen Anne Chateau   
4       Charming craftsman 3 bdm house   

                                             summary  \
0                                                NaN   
1  Chemically sensitive? We've removed the irrita...   
2  New modern house built in 2013.  Spectacular s...   
3  A charming apartment that sits atop Queen Anne...   
4  Cozy family craftman house in beautiful neighb...   

                                               space  \
0  Make your self at home in this charming one-be...   
1  Beautiful, hypoallergenic apartment in an extr...   
2  Our house is modern, light and fresh with a wa...   
3                                                NaN   
4  Cozy family craftman house in beautiful neighb...   

                                         description experiences_offered  \
0  Make your self at home in this charming one-be...                none   
1  Chemically sensitive? We've removed the irrita...                none   
2  New modern house built in 2013.  Spectacular s...                none   
3  A charming apartment that sits atop Queen Anne...                none   
4  Cozy family craftman house in beautiful neighb...                none   

                               neighborhood_overview  ... review_scores_value  \
0                                                NaN  ...                10.0   
1  Queen Anne is a wonderful, truly functional vi...  ...                10.0   
2  Upper Queen Anne is a charming neighborhood fu...  ...                10.0   
3                                                NaN  ...                 NaN   
4  We are in the beautiful neighborhood of Queen ...  ...                 9.0   

  requires_license license jurisdiction_names instant_bookable  \
0                f     NaN         WASHINGTON                f   
1                f     NaN         WASHINGTON                f   
2                f     NaN         WASHINGTON                f   
3                f     NaN         WASHINGTON                f   
4                f     NaN         WASHINGTON                f   

  cancellation_policy  require_guest_profile_picture  \
0            moderate                              f   
1              strict                              t   
2              strict                              f   
3            flexible                              f   
4              strict                              f   

  require_guest_phone_verification calculated_host_listings_count  \
0                                f                              2   
1                                t                              6   
2                                f                              2   
3                                f                              1   
4                                f                              1   

  reviews_per_month  
0              4.07  
1              1.48  
2              1.15  
3               NaN  
4              0.89  

[5 rows x 92 columns]

In [14]:
dfr.head()

listing_id        id        date  reviewer_id reviewer_name  \
0     7202016  38917982  2015-07-19     28943674        Bianca   
1     7202016  39087409  2015-07-20     32440555         Frank   
2     7202016  39820030  2015-07-26     37722850           Ian   
3     7202016  40813543  2015-08-02     33671805        George   
4     7202016  41986501  2015-08-10     34959538          Ming   

                                            comments  
0  Cute and cozy place. Perfect location to every...  
1  Kelly has a great room in a very central locat...  
2  Very spacious apartment, and in a great neighb...  
3  Close to Seattle Center and all it has to offe...  
4  Kelly was a great host and very accommodating ...

In [16]:
dfl = dfl.rename({'id': 'listing_id'}, axis=1)
dfl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 92 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   listing_id                        3818 non-null   int64  
 1   listing_url                       3818 non-null   object 
 2   scrape_id                         3818 non-null   int64  
 3   last_scraped                      3818 non-null   object 
 4   name                              3818 non-null   object 
 5   summary                           3641 non-null   object 
 6   space                             3249 non-null   object 
 7   description                       3818 non-null   object 
 8   experiences_offered               3818 non-null   object 
 9   neighborhood_overview             2786 non-null   object 
 10  notes                             2212 non-null   object 
 11  transit                           2884 non-null   object 
 12  thumbn

In [18]:
dfl = dfl[['listing_id','name','host_id','host_response_time','host_response_rate','host_acceptance_rate','host_is_superhost','host_identity_verified','city','state','zipcode','market','smart_location','country_code','latitude','longitude','is_location_exact','property_type','room_type','accommodates','bathrooms','bedrooms','beds','bed_type','price','weekly_price','monthly_price','guests_included','minimum_nights','maximum_nights','review_scores_rating','instant_bookable','cancellation_policy','require_guest_profile_picture','require_guest_phone_verification']]

In [20]:
dfl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3818 entries, 0 to 3817
Data columns (total 35 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   listing_id                        3818 non-null   int64  
 1   name                              3818 non-null   object 
 2   host_id                           3818 non-null   int64  
 3   host_response_time                3295 non-null   object 
 4   host_response_rate                3295 non-null   object 
 5   host_acceptance_rate              3045 non-null   object 
 6   host_is_superhost                 3816 non-null   object 
 7   host_identity_verified            3816 non-null   object 
 8   city                              3818 non-null   object 
 9   state                             3818 non-null   object 
 10  zipcode                           3811 non-null   object 
 11  market                            3818 non-null   object 
 12  smart_

In [22]:
dfc = dfc.rename({'price': 'payment'}, axis=1)
dfc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1393570 entries, 0 to 1393569
Data columns (total 4 columns):
 #   Column      Non-Null Count    Dtype 
---  ------      --------------    ----- 
 0   listing_id  1393570 non-null  int64 
 1   date        1393570 non-null  object
 2   available   1393570 non-null  object
 3   payment     934542 non-null   object
dtypes: int64(1), object(3)
memory usage: 42.5+ MB


In [26]:
# Menggabungkan DataFrame calendar dan listing berdasarkan 'listing_id'
df_merged = pd.merge(dfc, dfl, on='listing_id', how='left')
df_merged

listing_id        date available payment  \
0            241032  2016-01-04         t  $85.00   
1            241032  2016-01-05         t  $85.00   
2            241032  2016-01-06         f     NaN   
3            241032  2016-01-07         f     NaN   
4            241032  2016-01-08         f     NaN   
...             ...         ...       ...     ...   
1393565    10208623  2016-12-29         f     NaN   
1393566    10208623  2016-12-30         f     NaN   
1393567    10208623  2016-12-31         f     NaN   
1393568    10208623  2017-01-01         f     NaN   
1393569    10208623  2017-01-02         f     NaN   

                                 name   host_id  host_response_time  \
0        Stylish Queen Anne Apartment    956883  within a few hours   
1        Stylish Queen Anne Apartment    956883  within a few hours   
2        Stylish Queen Anne Apartment    956883  within a few hours   
3        Stylish Queen Anne Apartment    956883  within a few hours   
4        Stylish Queen Anne Apartment    956883  within a few hours   
...                               ...       ...                 ...   
1393565     Large Lakefront Apartment  14703116        within a day   
1393566     Large Lakefront Apartment  14703116        within a day   
1393567     Large Lakefront Apartment  14703116        within a day   
1393568     Large Lakefront Apartment  14703116        within a day   
1393569     Large Lakefront Apartment  14703116        within a day   

        host_response_rate host_acceptance_rate host_is_superhost  ...  \
0                      96%                 100%                 f  ...   
1                      96%                 100%                 f  ...   
2                      96%                 100%                 f  ...   
3                      96%                 100%                 f  ...   
4                      96%                 100%                 f  ...   
...                    ...                  ...               ...  ...   
1393565               100%                  NaN                 f  ...   
1393566               100%                  NaN                 f  ...   
1393567               100%                  NaN                 f  ...   
1393568               100%                  NaN                 f  ...   
1393569               100%                  NaN                 f  ...   

        weekly_price monthly_price guests_included minimum_nights  \
0                NaN           NaN               2              1   
1                NaN           NaN               2              1   
2                NaN           NaN               2              1   
3                NaN           NaN               2              1   
4                NaN           NaN               2              1   
...              ...           ...             ...            ...   
1393565          NaN           NaN               1              1   
1393566          NaN           NaN               1              1   
1393567          NaN           NaN               1              1   
1393568          NaN           NaN               1              1   
1393569          NaN           NaN               1              1   

        maximum_nights review_scores_rating instant_bookable  \
0                  365                 95.0                f   
1                  365                 95.0                f   
2                  365                 95.0                f   
3                  365                 95.0                f   
4                  365                 95.0                f   
...                ...                  ...              ...   
1393565           1125                  NaN                f   
1393566           1125                  NaN                f   
1393567           1125                  NaN                f   
1393568           1125                  NaN                f   
1393569           1125                  NaN                f   

         cancellation_policy  re

In [28]:
# Definisi fungsi untuk memperbarui status berdasarkan nilai Payment, Day Price, Weekly Price, Monthly Price

def update_status(df):
    """Fungsi untuk memperbarui kolom Day Price Status, Weekly Price Status, dan Monthly Price Status berdasarkan kolom Payment."""
    df['day_price_status'] = df.apply(lambda row: 'Yes' if row['payment'] == row['price'] else 'No', axis=1)
    df['weekly_price_status'] = df.apply(lambda row: 'Yes' if row['payment'] == row['weekly_price'] else 'No', axis=1)
    df['monthly_price_status'] = df.apply(lambda row: 'Yes' if row['payment'] == row['monthly_price'] else 'No', axis=1)
    return df

# Menggunakan fungsi update_status untuk memperbarui status pada DataFrame yang ada
df_result_updated = update_status(df_merged)
df_result_updated[['listing_id', 'date', 'payment', 'day_price_status', 'weekly_price_status', 'monthly_price_status']]

listing_id        date payment day_price_status weekly_price_status  \
0            241032  2016-01-04  $85.00              Yes                  No   
1            241032  2016-01-05  $85.00              Yes                  No   
2            241032  2016-01-06     NaN               No                  No   
3            241032  2016-01-07     NaN               No                  No   
4            241032  2016-01-08     NaN               No                  No   
...             ...         ...     ...              ...                 ...   
1393565    10208623  2016-12-29     NaN               No                  No   
1393566    10208623  2016-12-30     NaN               No                  No   
1393567    10208623  2016-12-31     NaN               No                  No   
1393568    10208623  2017-01-01     NaN               No                  No   
1393569    10208623  2017-01-02     NaN               No                  No   

        monthly_price_status  
0                         No  
1                         No  
2                         No  
3                         No  
4                         No  
...                      ...  
1393565                   No  
1393566                   No  
1393567                   No  
1393568                   No  
1393569                   No  

[1393570 rows x 6 columns]

**pakai update_status setiap kali ada perubahan di kolom payment**

# Descriptive Statistic

In [30]:
df_result_updated.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1393570 entries, 0 to 1393569
Data columns (total 41 columns):
 #   Column                            Non-Null Count    Dtype  
---  ------                            --------------    -----  
 0   listing_id                        1393570 non-null  int64  
 1   date                              1393570 non-null  object 
 2   available                         1393570 non-null  object 
 3   payment                           934542 non-null   object 
 4   name                              1393570 non-null  object 
 5   host_id                           1393570 non-null  int64  
 6   host_response_time                1202675 non-null  object 
 7   host_response_rate                1202675 non-null  object 
 8   host_acceptance_rate              1111425 non-null  object 
 9   host_is_superhost                 1392840 non-null  object 
 10  host_identity_verified            1392840 non-null  object 
 11  city                              139

In [32]:
# Menyaring baris dengan kondisi day_price_status, weekly_price_status, monthly_price_status sama-sama "No"
filtered_df = df_result_updated[(df_result_updated['day_price_status'] == 'No') & 
                                (df_result_updated['weekly_price_status'] == 'No') &
                                (df_result_updated['monthly_price_status'] == 'No')]

# Menghitung jumlah unik listing_id dalam hasil filter
unique_listing_count = filtered_df['listing_id'].nunique()

unique_listing_count

3423

Hal ini berarti ada 3423 listing yang ada ketidakcocokan payment dengan kolom-kolom status disuatu tanggal.

In [34]:
unique_date_count = filtered_df['date'].nunique()

unique_date_count

365

In [36]:
filtered_df

listing_id        date available payment  \
2            241032  2016-01-06         f     NaN   
3            241032  2016-01-07         f     NaN   
4            241032  2016-01-08         f     NaN   
5            241032  2016-01-09         f     NaN   
6            241032  2016-01-10         f     NaN   
...             ...         ...       ...     ...   
1393565    10208623  2016-12-29         f     NaN   
1393566    10208623  2016-12-30         f     NaN   
1393567    10208623  2016-12-31         f     NaN   
1393568    10208623  2017-01-01         f     NaN   
1393569    10208623  2017-01-02         f     NaN   

                                 name   host_id  host_response_time  \
2        Stylish Queen Anne Apartment    956883  within a few hours   
3        Stylish Queen Anne Apartment    956883  within a few hours   
4        Stylish Queen Anne Apartment    956883  within a few hours   
5        Stylish Queen Anne Apartment    956883  within a few hours   
6        Stylish Queen Anne Apartment    956883  within a few hours   
...                               ...       ...                 ...   
1393565     Large Lakefront Apartment  14703116        within a day   
1393566     Large Lakefront Apartment  14703116        within a day   
1393567     Large Lakefront Apartment  14703116        within a day   
1393568     Large Lakefront Apartment  14703116        within a day   
1393569     Large Lakefront Apartment  14703116        within a day   

        host_response_rate host_acceptance_rate host_is_superhost  ...  \
2                      96%                 100%                 f  ...   
3                      96%                 100%                 f  ...   
4                      96%                 100%                 f  ...   
5                      96%                 100%                 f  ...   
6                      96%                 100%                 f  ...   
...                    ...                  ...               ...  ...   
1393565               100%                  NaN                 f  ...   
1393566               100%                  NaN                 f  ...   
1393567               100%                  NaN                 f  ...   
1393568               100%                  NaN                 f  ...   
1393569               100%                  NaN                 f  ...   

        minimum_nights maximum_nights review_scores_rating instant_bookable  \
2                    1            365                 95.0                f   
3                    1            365                 95.0                f   
4                    1            365                 95.0                f   
5                    1            365                 95.0                f   
6                    1            365                 95.0                f   
...                ...            ...                  ...              ...   
1393565              1           1125                  NaN                f   
1393566              1           1125                  NaN                f   
1393567              1           1125                  NaN                f   
1393568              1           1125                  NaN                f   
1393569              1           1125                  NaN                f   

        cancellation_policy require_guest_profile_picture  \
2                  moderate                             f   
3                  moderate                             f   
4                  moderate                             f   
5                  moderate                             f   
6                  moderate                             f   
...                     ...                           ...   
1393565            flexible                             f   
1393566            flexible                             f   
1393567            flexible                             f   
1393568            flexible                             f   
1393569   

rows semuanya ada 1.393.570. rows yang tidak ada status nya 711.094. Ada 682.476 rows dari semua listing di seluruh tanggal yang sesuai dengan statusnya.

In [39]:
payment_no_status_nan = filtered_df['payment'].isna().sum()
print(payment_no_status_nan)

459028


Dari 711.094 rows yang tidak ada status, 459.028 rows tidak punya angkanya.

In [42]:
df_result_updated['payment'] = df_result_updated['payment'].str.replace('$', '').str.replace(',', '').astype(float)
df_result_updated['weekly_price'] = df_result_updated['weekly_price'].str.replace('$', '').str.replace(',', '').astype(float)
df_result_updated['monthly_price'] = df_result_updated['monthly_price'].str.replace('$', '').str.replace(',', '').astype(float)
df_result_updated['price'] = df_result_updated['price'].str.replace('$', '').str.replace(',', '').astype(float)

In [46]:
df_result_updated.head()

listing_id        date available  payment                          name  \
0      241032  2016-01-04         t     85.0  Stylish Queen Anne Apartment   
1      241032  2016-01-05         t     85.0  Stylish Queen Anne Apartment   
2      241032  2016-01-06         f      NaN  Stylish Queen Anne Apartment   
3      241032  2016-01-07         f      NaN  Stylish Queen Anne Apartment   
4      241032  2016-01-08         f      NaN  Stylish Queen Anne Apartment   

   host_id  host_response_time host_response_rate host_acceptance_rate  \
0   956883  within a few hours                96%                 100%   
1   956883  within a few hours                96%                 100%   
2   956883  within a few hours                96%                 100%   
3   956883  within a few hours                96%                 100%   
4   956883  within a few hours                96%                 100%   

  host_is_superhost  ... minimum_nights maximum_nights review_scores_rating  \
0                 f  ...              1            365                 95.0   
1                 f  ...              1            365                 95.0   
2                 f  ...              1            365                 95.0   
3                 f  ...              1            365                 95.0   
4                 f  ...              1            365                 95.0   

  instant_bookable cancellation_policy require_guest_profile_picture  \
0                f            moderate                             f   
1                f            moderate                             f   
2                f            moderate                             f   
3                f            moderate                             f   
4                f            moderate                             f   

  require_guest_phone_verification  day_price_status  weekly_price_status  \
0                                f               Yes                   No   
1                                f               Yes                   No   
2                                f                No                   No   
3                                f                No                   No   
4                                f                No                   No   

  monthly_price_status  
0                   No  
1                   No  
2                   No  
3                   No  
4                   No  

[5 rows x 41 columns]

In [52]:
df_result_updated['latitude'] = df_result_updated['latitude'].astype(str)
df_result_updated['longitude'] = df_result_updated['longitude'].astype(str)

In [54]:
df_result_updated.isna().sum()

listing_id                               0
date                                     0
available                                0
payment                             459028
name                                     0
host_id                                  0
host_response_time                  190895
host_response_rate                  190895
host_acceptance_rate                282145
host_is_superhost                      730
host_identity_verified                 730
city                                     0
state                                    0
zipcode                               2555
market                                   0
smart_location                           0
country_code                             0
latitude                                 0
longitude                                0
is_location_exact                        0
property_type                          365
room_type                                0
accommodates                             0
bathrooms  

In [56]:
df_result_updated.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1393570 entries, 0 to 1393569
Data columns (total 41 columns):
 #   Column                            Non-Null Count    Dtype  
---  ------                            --------------    -----  
 0   listing_id                        1393570 non-null  int64  
 1   date                              1393570 non-null  object 
 2   available                         1393570 non-null  object 
 3   payment                           934542 non-null   float64
 4   name                              1393570 non-null  object 
 5   host_id                           1393570 non-null  int64  
 6   host_response_time                1202675 non-null  object 
 7   host_response_rate                1202675 non-null  object 
 8   host_acceptance_rate              1111425 non-null  object 
 9   host_is_superhost                 1392840 non-null  object 
 10  host_identity_verified            1392840 non-null  object 
 11  city                              139

In [58]:
df_result_updated.describe()

listing_id        payment       host_id  accommodates     bathrooms  \
count  1.393570e+06  934542.000000  1.393570e+06  1.393570e+06  1.387730e+06   
mean   5.550111e+06     137.944859  1.578556e+07  3.349398e+00  1.259469e+00   
std    2.962274e+06     105.062870  1.458191e+07  1.977341e+00  5.902917e-01   
min    3.335000e+03      10.000000  4.193000e+03  1.000000e+00  0.000000e+00   
25%    3.258213e+06      75.000000  3.271389e+06  2.000000e+00  1.000000e+00   
50%    6.118244e+06     109.000000  1.055814e+07  3.000000e+00  1.000000e+00   
75%    8.035212e+06     160.000000  2.590413e+07  4.000000e+00  1.000000e+00   
max    1.034016e+07    1650.000000  5.320861e+07  1.600000e+01  8.000000e+00   

           bedrooms          beds         price   weekly_price  monthly_price  \
count  1.391380e+06  1.393205e+06  1.393570e+06  733285.000000  553705.000000   
mean   1.307712e+00  1.735394e+00  1.279762e+02     788.481832    2613.336190   
std    8.832794e-01  1.139331e+00  9.023823e+01     532.220123    1721.701786   
min    0.000000e+00  1.000000e+00  2.000000e+01     100.000000     500.000000   
25%    1.000000e+00  1.000000e+00  7.500000e+01     455.000000    1512.000000   
50%    1.000000e+00  1.000000e+00  1.000000e+02     650.000000    2200.000000   
75%    2.000000e+00  2.000000e+00  1.500000e+02     950.000000    3150.000000   
max    7.000000e+00  1.500000e+01  1.000000e+03    6300.000000   19500.000000   

       guests_included  minimum_nights  maximum_nights  review_scores_rating  
count     1.393570e+06    1.393570e+06    1.393570e+06          1.157415e+06  
mean      1.672603e+00    2.369303e+00    7.804476e+02          9.453926e+01  
std       1.310868e+00    1.630377e+01    1.683369e+03          6.605044e+00  
min       0.000000e+00    1.000000e+00    1.000000e+00          2.000000e+01  
25%       1.000000e+00    1.000000e+00    6.000000e+01          9.300000e+01  
50%       1.000000e+00    2.000000e+00    1.125000e+03          9.600000e+01  
75%       2.000000e+00    2.000000e+00    1.125000e+03          9.900000e+01  
max       1.500000e+01    1.000000e+03    1.000000e+05          1.000000e+02